### PR TITLE
Handle orphans with hashtable instead of list

### PR DESCRIPTION
**Pull request by sgaragagghu**
Loading from disk speedup is around 400%

define on compiler USE_BASIC_CONFIG if you can't run the configure of secp256k1 (like Windows OS) to create the configuration of secp256k1

### CODE REVIEW

**Comment by sgaragagghu** [comment]
Finish

**Comment by sgaragagghu** [comment]
Finish

File: client/block.c [hunk 1]
```diff
@@ -95,13 +105,21 @@ static pthread_mutex_t block_mutex;
 //TODO: this variable duplicates existing global variable g_is_pool. Probably should be removed
 static int g_light_mode = 0;
 static uint32_t cache_bounded_counter = 0;
+static struct orphan_block **orphan_hashtable = NULL;
```
Comment (xrdavies) [inline client/block.c]: When I was reviewing this PR, I checked again and again about this global variable. I suggest name it as a global variable.

File: client/block.c [hunk 2]
```diff
@@ -1038,6 +1040,11 @@ static void *work_thread(void *arg)
 	int n_mining_threads = (int)(unsigned)(uintptr_t)arg, sync_thread_running = 0;
 	uint64_t nhashes0 = 0, nhashes = 0;
 	pthread_t th;
+	if(USE_ORPHAN_HASHTABLE){
```
Comment (xrdavies) [inline client/block.c]: This macro can be removed since it's not used in the rest of code changes.

File: client/block.c (hunk 1 above)
Comment (sgaragagghu) [reply client/block.c]: Every block related variable in block.c isn't named as global

File: client/block.c (hunk 2 above)
Comment (sgaragagghu) [reply client/block.c]: But here is used ! :)

Comment (xrdavies) [reply client/block.c]: If USE_ORPHAN_HASHTABLE equals 0, what will happen?

Comment (sgaragagghu) [reply client/block.c]: It will use the original algorithm :D

**Comment by sgaragagghu** [comment]
define on compiler USE_BASIC_CONFIG if you can't run the configure of secp256k1 (like Windows OS) to create the configuration of secp256k1

File: client/block.c [hunk 3]
```diff
@@ -1771,3 +1782,110 @@ int xdag_get_transactions(xdag_hash_t hash, void *data, int (*callback)(void*, i
...
+		if(obt == NULL){
+			xdag_warn("Critical error. The orphan is not found in the hashtable list. [function: remove_orphan_hashtable]");
+			g_xdag_extstats.use_orphan_hashtable = 0;
+			clean_orphan_hashtable();
+		} else {
+			*(obt_back ? &obt_back->next_hashtable : obt_list_first) = obt->next_hashtable;
+
+			*blockRef0 = (obt->prev ? obt->prev->orphan_bi : NULL); // original list prev
+
+			*(obt->prev ? &((obt->prev)->next) : &orphan_first) = obt->next;
```
Comment (jonano614) [inline client/block.c]: Please, mark variables orphan_first and orphan_last as global.

File: client/block.c (hunk 1 above)
Comment (jonano614) [reply client/block.c]: Is it hashtable? Looks like array

Comment (sgaragagghu) [reply client/block.c]: it is an hashtable where buckets are lists

Comment (sgaragagghu) [reply client/block.c]: @jonano614 global variable original xdag logic is that, g_ variable are visible in all xdag not just in one file, by this logic those variable shouldn't have g_. Do you still want g_ on these variables ?

File: client/block.c [hunk 4]
```diff
@@ -1771,3 +1782,110 @@ int xdag_get_transactions(xdag_hash_t hash, void *data, int (*callback)(void*, i
...
 }
+
+void remove_orphan(struct block_internal* bi, struct block_internal* blockRef, struct block_internal* blockRef0){
+	if(!(bi->flags & BI_REF)) {
+		if(g_xdag_extstats.use_orphan_hashtable){
+			remove_orphan_hashtable(bi, &blockRef, &blockRef0);
+		}
+
+		// original remove orphan
+		if(!g_xdag_extstats.use_orphan_hashtable){
```
[the remark this answers is not in the document]
Comment (sgaragagghu) [reply client/block.c]: if and error occurs in remove_orphan_hashtable(bi, &blockRef, &blockRef0), it  will change g_xdag_extstats.use_orphan_hashtable so will do the original algorithm.
that's why there isn't if else.

File: client/block.c [hunk 5]
```diff
@@ -1771,3 +1782,110 @@ int xdag_get_transactions(xdag_hash_t hash, void *data, int (*callback)(void*, i
...
+		}
+
+		blockRef->ref = 0;
+		bi->flags |= BI_REF;
+		g_xdag_extstats.nnoref--;
+	}
+}
+
+void remove_orphan_hashtable(struct block_internal* bi, struct block_internal** blockRef, struct block_internal** blockRef0){
+	struct orphan_block **obt_list_first;
```
[the remark this answers is not in the document]
Comment (sgaragagghu) [reply client/block.c]: no

File: client/block.c [hunk 6]
```diff
@@ -1771,3 +1782,110 @@ int xdag_get_transactions(xdag_hash_t hash, void *data, int (*callback)(void*, i
...
+	obt_list_first = get_orphan_list(bi->hash);
+	struct orphan_block *obt = NULL;
+	if((obt = *obt_list_first) == NULL){
```
[the remark this answers is not in the document]
Comment (sgaragagghu) [reply client/block.c]: why?

File: client/block.c [hunk 7]
```diff
@@ -1771,3 +1782,110 @@ int xdag_get_transactions(xdag_hash_t hash, void *data, int (*callback)(void*, i
...
+		xdag_warn("Critical error. List in the hashtable not found. The orphan is not found in hashtable. [function: remove_orphan_hashtable]");
+		g_xdag_extstats.use_orphan_hashtable = 0;
+		clean_orphan_hashtable();
+	} else {
+		struct orphan_block *obt_back = NULL;
+
+		for(;(obt != NULL ? obt->orphan_bi != bi : 0); obt_back=obt, obt=obt->next_hashtable);
```
[the remark this answers is not in the document]
Comment (sgaragagghu) [reply client/block.c]: oh yes that's stupid.. i think i was tires when i did it

File: client/block.c [hunk 8]
```diff
@@ -450,7 +468,7 @@ static int add_block_nolock(struct xdag_block *newBlock, xdag_time_t limit)
 	int signinmask = 0, signoutmask = 0;
 	int inmask = 0, outmask = 0;
 	int verified_keys_mask = 0, err, type;
-	struct block_internal tmpNodeBlock, *blockRef, *blockRef0;
+	struct block_internal tmpNodeBlock, *blockRef = NULL, *blockRef0 = NULL;
```
[the remark this answers is not in the document]
Comment (sgaragagghu) [reply client/block.c]: it is a service variable, it has different meaning in every part of add_block_nolock, so i can change it with blockRef_temp, it hasn't only one meaning.
what do you think?

**Comment by jonano614** [comment]
@sgaragagghu yes, all global variables have to have prefix "g_"

**Comment by sgaragagghu** [comment]
@jonano614 i added g_ to global variables

File: client/block.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (jonano614) [reply client/block.c]: yes, blockRef_temp is better. "_temp" gives information that this variable is for temporary purposes, "0" gives nothing.

**Comment by jonano614** [comment]
Thank you for your efforts!

**Pull request merged**